### ISSUE TITLE
Call API functions directly from vimscript

### ISSUE THREAD

**Issue by bfredl** [neovim/neovim#4091]
In analogy with #1446 it would be useful to be able to call API functions directly from vimscript in the same instance, so that new features don't need to define both API and vimscript interfaces. Some prior discussion in #1817.

@justinmk suggested we should figure out the right interface before starting work on implementation.

**Comment by tarruda** [comment]
How about following the same method that is currently used for exposing the API via msgpack-rpc? For example: Reuse msgpack-gen.lua to generate vimL functions that call the API and automatically convert to/from vimL data structures.  Not sure about naming conventions for the generated functions though.

A simpler(but inefficient) way is to use a special channel/id that references the current nvim instance, and allow this id as argument to functions like `rpcrequest` or `rpcnotify`(which would be the same as `rpcrequest` except it doesn't return a value).

**Comment by justinmk** [comment]
> Reuse msgpack-gen.lua to generate vimL functions that call the API and automatically convert to/from vimL data structures.

That's a second, nice-to-have step. The first step is to make it easy to connect to a nvim server. VIMRUNTIME should not be needed to call API functions from a nvim instance.

> A simpler(but inefficient) way is to use a special channel/id that references the current nvim instance

That wouldn't be necessary if we could connect to a nvim server from VimL, then `v:servername` could be connected to and API can be called from that connection.

**Comment by bfredl** [comment]
If we want this to be a primary approach for using new features, the user should not need to type a magic number and the full method name to a special `rpc` function.

> Reuse msgpack-gen.lua to generate vimL functions that call the API and automatically convert to/from vimL data structures.

Something like that, but how do we handle the impedance mismatch that vimL mostly operates on curwin/curbuf while the API is object oriented? Using dict-function syntax we could expose an OO interface in vimL, but do we want that?

**Comment by bfredl** [comment]
> VIMRUNTIME should not be needed to call API functions from a nvim instance.

How is VIMRUNTIME needed to connect to an instance?

**Comment by justinmk** [comment]
> How is VIMRUNTIME needed to connect to an instance?

I assume the generated VimL functions are _user_ functions (`foo#bar()`). If they are shipped with the core as non-user functions then we are augmenting VimL even more. That would also mean that the nvim client must match the version of the nvim server.

**Comment by justinmk** [comment]
> Using dict-function syntax we could expose an OO interface in vimL, but do we want that?

I'd prefer not to use dict-function syntax, it's a second-class citizen in VimL. 

> vimL mostly operates on curwin/curbuf while the API is object oriented?

Maybe I'm missing something, why can't the API objects (Window, Buffer, Tabpage) be represented as dumb VimL dicts which are passed to the API functions?

**Comment by bfredl** [comment]
Aha, you want to use the same interface also to call functions on an a remote instance, now I see what you mean. But wouldn't the versions need to match anyway, I mean the client will use VIMRUNTIME of the cilent?
.
I had `apicall("method",...)` in mind for a low level interface but I suppose we could define `rpcrequest(0,...)` to short-circuit and call the API of the current instance (without msgpack marshalling).

**Comment by tarruda** [comment]
> Something like that, but how do we handle the impedance mismatch that vimL mostly operates on curwin/curbuf while the API is object oriented? Using dict-function syntax we could expose an OO interface in vimL, but do we want that?

vimL operates on curbuf/curwin because most internal functions work like this, at least for now. When I first designed the API I wanted it to be future-proof for when we remove globals from the internal functions(Eventually I hope that the compatibility code will be implemented in vimL functions and not API functions).

As for how we should handle this, I think the generated functions should simply receive the buf/win/tab id as first argument which is how vimL can call functions from other instances.(pure vimL object-oriented wrappers can easily be built upon this)

**Comment by bfredl** [comment]
> dumb VimL dicts which are passed to the API functions?

That would be fine, but do we still want them to be polymophic (`api#get_option(mybuf,...)` ) or rather `api#buffer_get_option(...)` ?

**Comment by ZyX-I** [comment]
@justinmk One can represent anything you can send via msgpack with special dictionaries described in `msgpackparse()`, no need to invent new entities. Also note #3417 and specifically new Vim type.

**Comment by bfredl** [comment]
And, rather than generalizing with calling remote instances, I actually thought more of coordination with the potential plan of reimplementing the vimscript interface on top of the API.

**Comment by tarruda** [comment]
> That wouldn't be necessary if we could connect to a nvim server from VimL, then v:servername could be connected-to and API functions can be called from that connection.

There's no need to pass through the connection layer if calling rpc methods in the same instance. As I said, a special channel id could be used to simply write the msgpack data to a local buffer, which is then read/parsed by the same msgpack routines as if the data came from a connection. Also, connecting to the self would complicate due to deadlock: `rpcrequest` enters a recursive event loop until the call returns, so how we would handle the request?

**Comment by bfredl** [comment]
Why would we even need to involve msgpack? We could just convert between the vimL typesystem and the API typesystem directly. (`vim_to_object`)

**Comment by ZyX-I** [comment]
@tarruda Buffer/window/tabpage number can be used as function argument only if there is something generated. But `apicall()` suggested by @bfredl may be added without generating anything if it is the way to go.

Though when I saw this issue I imagined (generated) `api_…` VimL functions which directly call C functions without involving msgpack or usual API dispatching methods at all. This can use numbers.

**Comment by bfredl** [comment]
Also there is no need to "wait" or involve the event look or anything. Just call the api function synchronously :)

**Comment by tarruda** [comment]
> Why would we even need to involve msgpack? We could just convert between the vimL typesystem and the API typesystem directly. (vim_to_object)

The only reason to involve msgpack is to reuse the existing infrastructure. I can probably enable self-calling API functions through `rpcrequest` without much work(special channel id). If we were to bypass msgpack, then we'd need to modify msgpack-gen.lua to generate vimL wrappers.

**Comment by bfredl** [comment]
You mean the lookup table for api function names is too strongly coupled with msgpack right now?

**Comment by ZyX-I** [comment]
@tarruda I used to generate such C code for lua, and it worked in a very few tests I performed. But that was not finished.

@bfredl I did not see anything like this or I would not be able to easily generate lua wrappers.

**Comment by bfredl** [comment]
`build/src/auto/msgpack_dispatch.c` doesn't look directly coupled to msgpack marshalling except for its name.

**Comment by tarruda** [comment]
> You mean the lookup table for api function names is too strongly coupled with msgpack right now?

It's not coupled, But it is not clear how you plan to implement `api#` vimL functions, unless you want to enhance vimL syntax to check for the `api#` prefix? Can you give an example of how `api#get_option(mybuf,...)` would work?

**Comment by bfredl** [comment]
Sorry for the confusion, I just meant `apicall("method", ...)` should be pretty straightforward without involving msgpack (more than things that happen to be called "`_msgpack_`" right now)

**Comment by tarruda** [comment]
> Sorry for the confusion, I just meant apicall("method", ...) should be pretty straightforward without involving msgpack (more than things that happen to be called "_msgpack_" right now)

:+1: That would be fine and even simpler than reusing `rpcrequest`/rpcnotify`.

**Comment by ZyX-I** [comment]
@bfredl Note that vim_to_object is not what somebody wants to use for this case: it cannot generate booleans, nils, etc. One needs to extend it to understand special dictionaries, but use this functionality only for `apicall` (thins like EXT strings returned by e.g. `vim_eval` may be rather unexpected).

**Comment by bfredl** [comment]
`api#` was only because it was suggested right before that wrappers would be implemented in vimscript. To if high level interface should be implemented in nvim core or in VIMRUNTIME I'm not sure, but I lean somewhat to the former, as I would like this interface to feel built-in-and-natural  rather than something bolted on.

**Comment by bfredl** [comment]
@ZyX-I Surely extending `vim_to_object` to support `v:true` etc would be natural given this approach ?

**Comment by tarruda** [comment]
> To if high level interface should be implemented in nvim core or in VIMRUNTIME I'm not sure, but I lean somewhat to the former, as I would like this interface to feel built-in-and-natural rather than something bolted on.

One way to make these functions available without dependency on `VIMRUNTIME` is to simply embed the vimL code into the executable during build time, and execute it during nvim startup. We can even make the functions defined during startup phase undeletable so they feel more like builtins

**Comment by bfredl** [comment]
That is certainly built-in enough :+1:

**Comment by ZyX-I** [comment]
@tarruda I would surely welcome a built-in function which looks like a user function. If `api…` functions are built-in they should be `api_…`, not undeletable `api#…`. And I do not see any sense at all in writing this in VimL because generator has no problems creating C `f_api_…` functions.

**Comment by bfredl** [comment]
BTW our dispatcher is already OK with passing in (Integer)0/1 for a Boolean (and I suppose `f_api_` would do just the same)

**Comment by tarruda** [comment]
> @tarruda I would surely welcome a built-in function which looks like a user function. If api… functions are built-in they should be api_…, not undeletable api#…. And I do not see any sense at all in writing this in VimL because generator has no problems creating C f_api_… functions.

Sure. The "user-like" builtin functions are simply a solution to a previous @justinmk concern of relying too much on vimL code for implementing high levels API meant to be used by plugins, mostly due to the possibility of VIMRUNTIME not being available.

**Comment by justinmk** [comment]
> The "user-like" builtin functions are simply a solution to a previous @justinmk concern of relying too much on vimL code for implementing high levels API meant to be used by plugins,

Now that we have worked through it, prefixing generated VimL core functions with `api_*` seems like the best option: 

- for _UX_ and _back-compat_ purposes it is equivalent to having a single function named e.g. `apicall()`
- there are valid reasons for not marshaling over `rpcrequest`
- special-casing `api#` is a UX cost (surprising special cases) without benefit relative to `api_` prefix

**Comment by justinmk** [comment]
Closed by https://github.com/neovim/neovim/issues/4934
